source/common/common/callback_impl.h

```diff
@@ -61,8 +63,8 @@ template <typename... CallbackArgs> class CallbackManager {
                         [handle](const CallbackHolder& holder) -> bool {
                           return handle == &holder;
                         }) != callbacks_.end());
-    callbacks_.remove_if(
-        [handle](const CallbackHolder& holder) -> bool { return handle == &holder; });
+    auto it = static_cast<CallbackHolder*>(handle)->it_;
+    callbacks_.erase(it);
```

Comment:
IIUC this is still O(N) and I think the readability is not getting better. 
can you make callbacks_ a flat_hashmap as we've discussed, which could be a performance improvement here.